src/librbd/api/Utils.cc

```diff
@@ -74,6 +79,186 @@ int create_encryption_format(
   return 0;
 }
 
+template <typename I>
+int notify_quiesce(std::vector<I *> &ictxs, ProgressContext &prog_ctx,
+                   std::vector<uint64_t> *requests) {
+  int image_count = ictxs.size();
+  std::vector<C_SaferCond> on_finishes(image_count);
+
+  requests->resize(image_count);
+  for (int i = 0; i < image_count; ++i) {
+    auto ictx = ictxs[i];
+
+    ictx->image_watcher->notify_quiesce(&(*requests)[i], prog_ctx,
+                                        &on_finishes[i]);
+  }
+
+  int ret_code = 0;
+  for (int i = 0; i < image_count; ++i) {
+    int r = on_finishes[i].wait();
+    if (r < 0) {
+      ret_code = r;
+    }
+  }
+
+  if (ret_code != 0) {
+    notify_unquiesce(ictxs, *requests);
+  }
+
+  return ret_code;
+}
+
+template <typename I>
+void notify_unquiesce(std::vector<I *> &ictxs,
+                      const std::vector<uint64_t> &requests) {
+  if (requests.empty()) {
+    return;
+  }
+
+  ceph_assert(requests.size() == ictxs.size());
+  int image_count = ictxs.size();
+  std::vector<C_SaferCond> on_finishes(image_count);
+
+  for (int i = 0; i < image_count; ++i) {
+    ImageCtx *ictx = ictxs[i];
+
+    ictx->image_watcher->notify_unquiesce(requests[i], &on_finishes[i]);
+  }
+
+  for (int i = 0; i < image_count; ++i) {
+    on_finishes[i].wait();
+  }
+}
+
+template <typename I>
+librados::snap_t get_group_snap_id(
+    I *ictx, const cls::rbd::SnapshotNamespace& in_snap_namespace) {
+  ceph_assert(ceph_mutex_is_locked(ictx->image_lock));
+  auto it = ictx->snap_ids.lower_bound({cls::rbd::GroupImageSnapshotNamespace{},
+                                        ""});
+  for (; it != ictx->snap_ids.end(); ++it) {
+    if (it->first.first == in_snap_namespace) {
+      return it->second;
+    } else if (boost::get<cls::rbd::GroupImageSnapshotNamespace>(
+                   &it->first.first) == nullptr) {
+      break;
+    }
+  }
+  return CEPH_NOSNAP;
+}
+
+int group_snap_remove(librados::IoCtx& group_ioctx, const std::string& group_id,
+                      const cls::rbd::GroupSnapshot& group_snap) {
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  std::string group_header_oid = librbd::util::group_header_name(group_id);
+  std::vector<C_SaferCond*> on_finishes;
+  int r, ret_code;
+
+  std::vector<librbd::ImageCtx*> ictxs;
+
+  cls::rbd::GroupImageSnapshotNamespace ne{group_ioctx.get_id(), group_id,
```

Comment:
ne is not a descriptive variable name. It would be better to rename it to something like snap_namespace.
Holds for all functions.